module/VuFindSearch/src/VuFindSearch/Backend/Solr/LuceneSyntaxHelper.php

```diff
@@ -666,4 +665,63 @@ protected function capitalizeRangesCallback($match)
             return $open . trim($start) . ' TO ' . trim($end) . $close;
         }
     }
+
+    /**
+     * Count occurrences of a character in non-quoted parts of the string
+     *
+     * @param string $needle   Character to look for
+     * @param string $haystack String to process
+     *
+     * @return int
+     */
+    protected function countNonQuoted(string $needle, string $haystack): int
+    {
+        $count = 0;
+        $inQuotes = false;
+        $lastCh = '';
+        $lastLastCh = '';
+        foreach (str_split($haystack) as $ch) {
+            // Check for escaped character (i.e. preceding character is backslash
+            // that's not escaped):
+            if ('\\' !== $lastCh || '\\' === $lastLastCh) {
+                if ('"' === $ch) {
+                    $inQuotes = !$inQuotes;
+                }
+                if (!$inQuotes && $ch === $needle) {
+                    ++$count;
+                }
+            }
+            $lastLastCh = $lastCh;
+            $lastCh = $ch;
+        }
+        return $count;
+    }
+
+    /**
+     * Remove occurrences of given characters in non-quoted parts of the string
+     *
+     * @param array  $needles  Characters to remove
+     * @param string $haystack String to process
+     *
+     * @return string
+     */
+    protected function removeNonQuoted(array $needles, string $haystack): string
+    {
+        $result = '';
+        $inQuotes = false;
+        $lastCh = '';
+        $lastLastCh = '';
+        foreach (str_split($haystack) as $ch) {
+            // Check for non-escaped quote:
+            if ('"' === $ch && ('\\' !== $lastCh || '\\' === $lastLastCh)) {
```

Comment:
Is there a reason that the logic is a little bit different between countNonQuoted and removeNonQuoted? If not, I'm wondering if it might actually make sense to create a protected helper method so the non-quoted logic is only maintained in one place, and the rest of the work could be done with a callback... e.g., something like:

```
    protected function countNonQuoted(string $needle, string $haystack): int
    {
        $result = 0;
        $callback = function ($ch) use ($result) {
            $result++;
        };
        $this->applyCallbackToNonQuotedChars([$needle], $haystack, $callback);
        return $result;
    }

    protected function removeNonQuoted(array $needles, string $haystack): string
    {
        $result = '';
        $callback = function ($ch) use ($result) {
            $result .= $ch;
        };
        $this->applyCallbackToNonQuotedChars($needles, $haystack, $callback);
        return $result;
    }

    protected function applyCallbackToNonQuotedChars(array $needles, string $haystack, callable $callback): void
    {
        $inQuotes = false;
        $lastCh = '';
        $lastLastCh = '';
        foreach (str_split($haystack) as $ch) {
            // Check for non-escaped quote:
            if ('"' === $ch && ('\\' !== $lastCh || '\\' === $lastLastCh)) {
                $inQuotes = !$inQuotes;
            }
            if (!$inQuotes && !in_array($ch, $needles)) {
                $callback($ch);
            }
            $lastLastCh = $lastCh;
            $lastCh = $ch;
        }
    }
```

Too much abstraction? Maybe... but just trying to avoid replication of somewhat complex logic.

Reply:
Oops, my mistake. I'll try to find a way to avoid duplication. Also the escape handling is still not quite correct.